### PR TITLE
[SPARK-8625] [Core] Propagate user exceptions in tasks back to driver

### PR DESCRIPTION
**Pull request by tomwhite**
This allows clients to retrieve the original exception from the
cause field of the SparkException that is thrown by the driver.
If the original exception is not in fact Serializable then it will
not be returned, but the message and stacktrace will be. (All Java
Throwables implement the Serializable interface, but this is no
guarantee that a particular implementation can actually be
serialized.)

### CODE REVIEW

**Comment by sujkh85** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <[spark] [SPARK-8625] [Core] Propagate user exceptions in tasks back to driver (#7014)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by sujkh85** [comment]
NAVER - http://www.naver.com/
--------------------------------------------

sujkh@naver.com 님께 보내신 메일 <Re: [spark] [SPARK-8625] [Core] Propagate user exceptions in tasks back to driver (#7014)> 이 다음과 같은 이유로 전송 실패했습니다.

--------------------------------------------

받는 사람이 회원님의 메일을 수신차단 하였습니다. 


--------------------------------------------

**Comment by srowen** [comment]
OK to test

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 1]
```diff
@@ -745,8 +745,9 @@ class DAGScheduler(
     submitWaitingStages()
   }
 
-  private[scheduler] def handleTaskSetFailed(taskSet: TaskSet, reason: String) {
-    stageIdToStage.get(taskSet.stageId).foreach {abortStage(_, reason) }
+  private[scheduler] def handleTaskSetFailed(taskSet: TaskSet, reason: String,
+                                             exception: Option[Throwable] = None) {
+    stageIdToStage.get(taskSet.stageId).foreach {abortStage(_, reason, exception) }
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: nit:  multiline method defs should have each arg on a separate line, all indented 4 extra spaces.  Also include the return type (I know it wasn't there before but as long as you're touching this ...)

```scala
private[scheduler] def handleTaskSetFailed(
    taskSet: TaskSet,
    reason: String,
    exception: Option[Throwable] = None): Unit = {
```

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 2]
```diff
@@ -1155,7 +1156,8 @@ class DAGScheduler(
       case commitDenied: TaskCommitDenied =>
         // Do nothing here, left up to the TaskScheduler to decide how to handle denied commits
 
-      case ExceptionFailure(className, description, stackTrace, fullStackTrace, metrics) =>
+      case ExceptionFailure(className, description, stackTrace, fullStackTrace, metrics,
+                            exception) =>
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: nit: also not something you introduced, but while you're touching this, it can be cleaned up.  If you're not using any of those fields, we can just check the class in the case statement  `case _: ExceptionFailure =>`.  or since there are so many uses of underscore in scala and they can be kinda confusing, you can bind the whole thing to a variable (even though we'll never look at it): `case ef: ExceptionFailure =>`

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 3]
```diff
@@ -1264,7 +1266,8 @@ class DAGScheduler(
    * Aborts all jobs depending on a particular Stage. This is called in response to a task set
    * being canceled by the TaskScheduler. Use taskSetFailed() to inject this event from outside.
    */
-  private[scheduler] def abortStage(failedStage: Stage, reason: String) {
+  private[scheduler] def abortStage(failedStage: Stage, reason: String, exception:
+  Option[Throwable] = None) {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: nit: style for multiline & include return type

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 4]
```diff
@@ -1281,8 +1284,9 @@ class DAGScheduler(
   }
 
   /** Fails a job and all stages that are only used by that job, and cleans up relevant state. */
-  private def failJobAndIndependentStages(job: ActiveJob, failureReason: String) {
-    val error = new SparkException(failureReason)
+  private def failJobAndIndependentStages(job: ActiveJob, failureReason: String,
+                                          exception: Option[Throwable] = None) {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: nit: style for multiline & include return type

File: core/src/main/scala/org/apache/spark/scheduler/DAGSchedulerEvent.scala [hunk 5]
```diff
@@ -74,6 +74,7 @@ private[scheduler] case class ExecutorAdded(execId: String, host: String) extend
 private[scheduler] case class ExecutorLost(execId: String) extends DAGSchedulerEvent
 
 private[scheduler]
-case class TaskSetFailed(taskSet: TaskSet, reason: String) extends DAGSchedulerEvent
+case class TaskSetFailed(taskSet: TaskSet, reason: String,
+                         exception: Option[Throwable] = None) extends DAGSchedulerEvent
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGSchedulerEvent.scala]: nit: multiline style

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 6]
```diff
@@ -662,6 +662,7 @@ private[spark] class TaskSetManager(
 
     val failureReason = s"Lost task ${info.id} in stage ${taskSet.id} (TID $tid, ${info.host}): " +
       reason.asInstanceOf[TaskFailedReason].toErrorString
+    var failureException: Option[Throwable] = None
     reason match {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: you could avoid using a `var` by making the exception the result of the entire match block, eg.

```scala
val failureException: Option[Throwable] = reason match {
 ...
```

honestly I don't have a strong opinion ... its nice to use a `val` so you know the value is set by the `match` block, but OTOH the `match` block is so big that is might be a bit hard to follow what is being returned ...

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala [hunk 7]
```diff
@@ -728,16 +730,16 @@ private[spark] class TaskSetManager(
...
-          .format(index, taskSet.id, maxTaskFailures, failureReason))
+          .format(index, taskSet.id, maxTaskFailures, failureReason), failureException)
         return
       }
     }
     maybeFinishTaskSet()
   }
 
-  def abort(message: String): Unit = sched.synchronized {
+  def abort(message: String, exception: Option[Throwable] = None): Unit = sched.synchronized {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: Do we really need to have the default ` = None` for all of these cases?  This isn't so much about this particular method definition as for all of the changes.  It seems like you are supplying the exception (or `None`) in most cases.

If it really is needed that is fine, I am just asking b/c I know I have a habit of putting in a lot of defaults early during some refactoring just to get things compiling, but then when its all done I've put in far more than necessary.

File: core/src/test/scala/org/apache/spark/FailureSuite.scala [hunk 8]
```diff
@@ -139,5 +139,32 @@ class FailureSuite extends SparkFunSuite with LocalSparkContext {
...
+  test("failure cause is sent back to driver") {
+    sc = new SparkContext("local", "test")
+    val data = sc.makeRDD(1 to 3).map(x => { throw new IllegalStateException("oops"); (x,
+      x) }).groupByKey(3)
+    val thrown = intercept[SparkException] {
+      data.collect()
+    }
+    assert(thrown.getClass === classOf[SparkException])
+    assert(thrown.getCause.getClass === classOf[IllegalStateException])
+  }
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/FailureSuite.scala]: super minor, but can you also check the contents of the `IllegalStateException`?  I'm sure its fine in your implementation, but might as well add a check that we don't regress and somehow don't serialize the full exception

**Comment by squito** [comment]
@tomwhite this looks great!  thanks so much for working on this, I think it will be a really good addition.  I left some minor style comments in addition to the ones the automatic checker found, but overall this seems close.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 6 above)
Comment (tomwhite) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I used a `var` for the second reason you stated: that the `match` block is so big it would be hard to follow.

**Comment by tomwhite** [comment]
@squito thanks for the review! I've addressed your comments in a new commit.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 9]
```diff
@@ -17,7 +17,7 @@
 
 package org.apache.spark.executor
 
-import java.io.File
+import java.io.{NotSerializableException, File}
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: nit: order alphabetically `{File, NotSerializableException}`

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 10]
```diff
@@ -1264,7 +1267,10 @@ class DAGScheduler(
    * Aborts all jobs depending on a particular Stage. This is called in response to a task set
    * being canceled by the TaskScheduler. Use taskSetFailed() to inject this event from outside.
    */
-  private[scheduler] def abortStage(failedStage: Stage, reason: String) {
+  private[scheduler] def abortStage(
+      failedStage: Stage,
+      reason: String, exception:
+      Option[Throwable] = None): Unit = {
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: nit: each arg on a separate line

File: core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala [hunk 11]
```diff
@@ -47,7 +47,10 @@ class FakeDAGScheduler(sc: SparkContext, taskScheduler: FakeTaskScheduler)
 
   override def executorLost(execId: String) {}
 
-  override def taskSetFailed(taskSet: TaskSet, reason: String) {
+  override def taskSetFailed(
+      taskSet: TaskSet,
+      reason: String, exception:
+      Option[Throwable]): Unit = {
```
Comment (squito) [inline core/src/test/scala/org/apache/spark/scheduler/TaskSetManagerSuite.scala]: nit: each arg on a separate line

**Comment by squito** [comment]
a couple of minor comments, aside from that just waiting to see if the tests pass

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 6 above)
Comment (markhamstra) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: I'd prefer that you avoid the unnecessary mutability.  The `match` cases are all quite small with the exception of the `ExceptionFailure` case, which means that even with the `var` it's not all that readable or easy to figure out just where `failureException` is being set and that it is not being mutated elsewhere.  If you really want to shorten the `match` and make it more readable, I suggest pulling the logic of the `ExceptionFailure` case out into a helper function defined within `def handleFailedTask`, then make each of the cases in the `match` produce `None` or `Some(throwable)`.

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala [hunk 12]
```diff
@@ -97,11 +101,17 @@ case class ExceptionFailure(
     description: String,
     stackTrace: Array[StackTraceElement],
     fullStackTrace: String,
-    metrics: Option[TaskMetrics])
+    metrics: Option[TaskMetrics],
+    exception: Option[Throwable] = None)
```
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: Will this break pattern matching on this class? Is it possible to add an `unapply` method to preserve matching for the old signature? I'm rusty on the details of backwards compatibility for case classes.

File: core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala (hunk 6 above)
Comment (tomwhite) [inline core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala]: @markhamstra Thanks for the comment. I had another look at made the change you suggested.

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala (hunk 12 above)
Comment (tomwhite) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: @pwendell You're right - it will break pattern matching on the class. My understanding is that an `unapply` method won't help since pattern matching won't use it (they are for user code).

Case classes don't play well with binary compatibility, it seems. To do this compatibly, we'd have to have another case class, called `ExceptionFailureWithCause` say, and a trait that both it and `ExceptionFailure` extend with the common fields. Then everywhere that handles `ExceptionFailure` would also have to handle `ExceptionFailureWithCause`.

Having said all that, this class is marked `@DeveloperApi` so it's within the contract to change it. The `fullStackTrace` field was added last November, for example. I can understand the general reluctance to change code even if it is marked as being for developers only, but it's not clear if the workaround here to preserve binary compatibility is worth the complexity it adds.

Comment (squito) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: I think we're definitely out of luck for binary compatibility, but I think @pwendell just wanted to preserve source compatibility (ie. maybe users will need to recompile, but they won't need to change their code at all).

However, I don't think that is possible either.  You would need to have another method like `def unapply(ef: ExceptionFailure): Option[(String, String, Array[StackTraceElement], String, Option[TaskMetrics])]` -- ie., exactly the same as the built-in unapply, but without the final `Option[Throwable]` in the return type.  But that isn't legal overloading -- it has the same set of arguments as the built-in `unapply`, just a different return type.

Is there another way around this I'm not seeing?  I agree we shouldn't change things willy-nilly just b/c its `@DeveloperApi`, but IMO this change is worth it.

**Comment by squito** [comment]
@pwendell are you OK with the change to `ExceptionFailure`?   If you're ok w/ it, this lgtm

**Comment by squito** [comment]
ping @pwendell

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala (hunk 12 above)
Comment (pwendell) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: Okay - if it's not possible to preserve matching, let's just change it. Adding fields will only break matching and shouldn't hut users who just access the fields. We might later on add some docs to suggest avoiding use of matching on developer API's for this reason (no need to block this though).

**Comment by pwendell** [comment]
I think the compatibility is okay, but two other quick questions:

1. Is it well defined which exception caused the task to fail? What if a task fails N times with N different exceptions, which is present in the cause? I think this should be documented clearly in the docs for the developer API.
2. Maybe @kayousterhout could look quickly at the DAG scheduler changes.

**Comment by markhamstra** [comment]
@pwendell I'm not seeing anything concerning in the DAGScheduler changes.

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala [hunk 13]
```diff
@@ -90,6 +90,10 @@ case class FetchFailed(
  *
  * `fullStackTrace` is a better representation of the stack trace because it contains the whole
  * stack trace including the exception and its causes
+ *
+ * `preserveCause` is used to keep the exception itself so it is available to the
+ * driver. This may be set to `false` in the event that the exception is not in fact
+ * serializable.
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: I think this is the wrong place for this doc on `preserveCause`.  That should go on the constructor that has it.  Here I'd just doc the `exception` (and add an explanation of which exception is corresponds to as Patrick asked)

**Comment by squito** [comment]
I played with this locally a bit, and I think that actually defaulting `Option[Throwable] = None` is covering up some cases where there really is an exception you should include.  I think you should just remove that defaulting everywhere.  It'll lead to a handful of compile errors, but mostly they will be easy to fix -- either there will be an obvious exception to pass or there won't ...

... except for the default on `ExceptionFailure`.  That will lead to a failure in [`JsonProtocol.taskEndReasonFromJson`](https://github.com/apache/spark/blob/master/core/src/main/scala/org/apache/spark/util/JsonProtocol.scala#L764).  Is it possible to turn the actual exception into json in `JsonProtocol.taskEndReasonToJson`, so that it can be read back in?  I was wondering if getting the full exception back might be useful to anything which wants to analyze the event logs.  I guess its impossible for some general purpose job to know about whatever user exceptions there are, so maybe there isn't any good reason for getting the full exception in json anyway.

So I'm leaning towards just putting in `None` for the exception in `JsonProtocol.taskEndReasonFromJson`, but just wanted to share my thoughts.

File: core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala [hunk 14]
```diff
@@ -745,8 +745,11 @@ class DAGScheduler(
     submitWaitingStages()
   }
 
-  private[scheduler] def handleTaskSetFailed(taskSet: TaskSet, reason: String) {
-    stageIdToStage.get(taskSet.stageId).foreach {abortStage(_, reason) }
+  private[scheduler] def handleTaskSetFailed(
+      taskSet: TaskSet,
+      reason: String,
+      exception: Option[Throwable]): Unit = {
+    stageIdToStage.get(taskSet.stageId).foreach {abortStage(_, reason, exception) }
```
Comment (kayousterhout) [inline core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala]: nit: space after "{" (or use parens instead, with no spaces around them)

**Comment by kayousterhout** [comment]
Scheduler changes LGTM, subject to @squito's suggestion

**Comment by tomwhite** [comment]
Thanks for all the feedback @pwendell, @squito, and @kayousterhout. I've addressed all your points and updated this PR.

@squito Regarding `JsonProtocol`, I can't see a way of serializing the fields of an arbitrary user exception class to JSON and the reconstituting them, so I agree that putting it as `None` is the right way to go. The stack trace and message will be preserved in the same way that they currently are though.

File: core/src/main/scala/org/apache/spark/executor/Executor.scala [hunk 15]
```diff
@@ -300,8 +300,16 @@ private[spark] class Executor(
               m
             }
           }
-          val taskEndReason = new ExceptionFailure(t, metrics)
-          execBackend.statusUpdate(taskId, TaskState.FAILED, ser.serialize(taskEndReason))
+          val serializedTaskEndReason = {
+            try {
+              ser.serialize(new ExceptionFailure(t, metrics))
```
Comment (aarondav) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: Is it possible that an exception does not throw a NotSerializableException but cannot be deserialized on the other side? This could be due to, say, class loader issues or other issues not caught during serialization time.

Comment (aarondav) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: (By the way, the particular exception I'm thinking of is Scala's MatchError, which actually includes the object which failed to match.)

Comment (tomwhite) [inline core/src/main/scala/org/apache/spark/executor/Executor.scala]: This is definitely possible. I've written a test for this case, and implemented a fix which uses a wrapper for the exception that gracefully falls back if deserialization of the exception fails.

**Comment by tomwhite** [comment]
The build was failing because the changes didn't compile against master. I've rebased and fixed the problem, as well as added a commit to address @aarondav's comment.

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala [hunk 16]
```diff
@@ -127,6 +146,19 @@ case class ExceptionFailure(
...
 
+class ThrowableSerializationWrapper(var exception: Throwable) extends Serializable {
+  private def writeObject(out: ObjectOutputStream): Unit = {
+    out.writeObject(exception)
+  }
+  private def readObject(in: ObjectInputStream): Unit = {
+    try {
+      exception = in.readObject().asInstanceOf[Throwable]
+    } catch {
+      case _ : ClassNotFoundException | _ : IOException => // exception is null
```
Comment (aarondav) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: We could log the error at warning level so users know why they're not seeing the actual exception.

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala [hunk 17]
```diff
@@ -97,13 +103,26 @@ case class ExceptionFailure(
...
+   * serializable.
+   */
+  private[spark] def this(e: Throwable, metrics: Option[TaskMetrics], preserveCause: Boolean) {
+    this(e.getClass.getName, e.getMessage, e.getStackTrace, Utils.exceptionString(e), metrics,
+      if (preserveCause) Some(new ThrowableSerializationWrapper(e)) else None)
+  }
+
   private[spark] def this(e: Throwable, metrics: Option[TaskMetrics]) {
-    this(e.getClass.getName, e.getMessage, e.getStackTrace, Utils.exceptionString(e), metrics)
+    this(e, metrics, true)
```
Comment (aarondav) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: perhaps used named parameter, `preservesCause = true`

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala [hunk 18]
```diff
@@ -97,13 +103,26 @@ case class ExceptionFailure(
...
   }
 
+  def exception = exceptionWrapper.map(_.exception)
```
Comment (aarondav) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: I think this would actually be `Some(null)` in the exceptional case, no?

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala [hunk 19]
```diff
@@ -127,6 +146,19 @@ case class ExceptionFailure(
   }
 }
 
+class ThrowableSerializationWrapper(var exception: Throwable) extends Serializable {
```
Comment (aarondav) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: Perhaps add a brief comment on the intention of this class (to avoid errors due to unserializable exceptions), and maybe indicate that exception will be null in this case.

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala [hunk 20]
```diff
@@ -97,13 +103,26 @@ case class ExceptionFailure(
     description: String,
     stackTrace: Array[StackTraceElement],
     fullStackTrace: String,
-    metrics: Option[TaskMetrics])
+    metrics: Option[TaskMetrics],
+    exceptionWrapper: Option[ThrowableSerializationWrapper])
```
Comment (aarondav) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: Would you mind adding a test for this guy? Could make a new Exception subtype which throws an exception in readObject() :)

Comment (tomwhite) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: Agreed - there's a test I added in FailureSuite that uses NonDeserializableUserException to check this exact case.

**Comment by tomwhite** [comment]
Thanks @aarondav for the review. I've addressed all your points and I also rebased on master to fix a conflict.

File: core/src/test/scala/org/apache/spark/FailureSuite.scala [hunk 21]
```diff
@@ -141,5 +141,73 @@ class FailureSuite extends SparkFunSuite with LocalSparkContext {
...
+class UserException(message: String, cause: Throwable)
+  extends RuntimeException(message, cause)
+
+class NonSerializableUserException extends RuntimeException {
+  val nonSerializableInstanceVariable = new NonSerializable
+}
+
+class NonDeserializableUserException extends RuntimeException {
+  private def readObject(in: ObjectInputStream): Unit = {
+    throw new IOException("Intentional exception during deserialization.");
```
Comment (aarondav) [inline core/src/test/scala/org/apache/spark/FailureSuite.scala]: nit: semicolon

File: core/src/test/scala/org/apache/spark/FailureSuite.scala [hunk 22]
```diff
@@ -141,5 +141,73 @@ class FailureSuite extends SparkFunSuite with LocalSparkContext {
...
+    }
+    assert(thrown.getClass === classOf[SparkException])
+    assert(thrown.getCause === null)
+    assert(thrown.getMessage.contains("NonSerializableUserException"))
+    FailureSuiteState.clear()
+  }
+
+  test("failure cause stacktrace is sent back to driver if exception is not deserializable") {
+    sc = new SparkContext("local", "test")
+    val data = sc.makeRDD(1 to 3).map(x => { throw new NonDeserializableUserException; (x,
```
Comment (aarondav) [inline core/src/test/scala/org/apache/spark/FailureSuite.scala]: nit: maybe just
```scala
val data = sc.makeRDD(1 to 3).foreach { _ =>
  throw new NonDeserializationUserException
}.groupByKey(3)
```

File: core/src/test/scala/org/apache/spark/FailureSuite.scala (hunk 21 above)
Comment (tomwhite) [inline core/src/test/scala/org/apache/spark/FailureSuite.scala]: Fixed

File: core/src/test/scala/org/apache/spark/FailureSuite.scala (hunk 22 above)
Comment (tomwhite) [inline core/src/test/scala/org/apache/spark/FailureSuite.scala]: Unless I'm missing something I don't think this works as `foreach` doesn't return values so there's not way to then to call `groupByKey`.

Comment (aarondav) [inline core/src/test/scala/org/apache/spark/FailureSuite.scala]: haha, whoops; that said, the groupByKey and collect should be unnecessary as the foreach should throw the exception directly

Comment (tomwhite) [inline core/src/test/scala/org/apache/spark/FailureSuite.scala]: @aarondav I was following the pattern of the other tests in the same suite, so perhaps it's OK to leave it like this?

Comment (squito) [inline core/src/test/scala/org/apache/spark/FailureSuite.scala]: I agree with aaron, I'd prefer they were more minimal.  The other tests with groupBy are trying to test shuffles in particular.  If somebody is reading this in isolation, its not clear what the point of the groupBy is.

(There is one of the tests w/ a `map` then `collect()` could probably be simplified to just a `foreach`, but oh well.)

Comment (tomwhite) [inline core/src/test/scala/org/apache/spark/FailureSuite.scala]: I've simplified this now.

**Comment by tomwhite** [comment]
Rebased on master to fix a merge conflict.

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala [hunk 23]
```diff
@@ -97,13 +103,27 @@ case class ExceptionFailure(
...
+      if (preserveCause) Some(new ThrowableSerializationWrapper(e)) else None)
+  }
+
   private[spark] def this(e: Throwable, metrics: Option[TaskMetrics]) {
-    this(e.getClass.getName, e.getMessage, e.getStackTrace, Utils.exceptionString(e), metrics)
+    this(e, metrics, preserveCause = true)
   }
 
+  def exception = exceptionWrapper.flatMap(
+    (w : ThrowableSerializationWrapper) => Option(w.exception))
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: nit: needs a return type, and if multiline wrap in braces.  Though you could also do `exceptionWrapper.flatMap(Option(_.exception))` which might eliminate the need for multiple lines

**Comment by tomwhite** [comment]
Looks like this change is not causing the failure in SQLQuerySuite as it is failing for other PRs (e.g. https://github.com/apache/spark/pull/7678, https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/1205/).

**Comment by squito** [comment]
Jenkins, retest this please

File: core/src/main/scala/org/apache/spark/TaskEndReason.scala [hunk 24]
```diff
@@ -128,6 +149,25 @@ case class ExceptionFailure(
 }
 
 /**
+ * A class for recovering from exceptions when deserializing a Throwable that was
+ * thrown in user task code. If the Throwable cannot be deserialized it will be null,
+ * but the stacktrace and message will be preserved correctly in SparkException.
+ */
+private[spark] class ThrowableSerializationWrapper(var exception: Throwable) extends
```
Comment (squito) [inline core/src/main/scala/org/apache/spark/TaskEndReason.scala]: its a good idea to make this `private[spark]` if we can -- but its actually going to leak as a member of `ExceptionFailure`.  (I'm surprised that scalac doesn't catch this.)  In any case, I think you can fix this by changing to the member of `ExceptionFailure` to `private val exceptionWrapper: Option[ThrowableSerializationWrapper]`.

**Comment by squito** [comment]
@aarondav are you OK with this now?  I think tom addressed all your concerns

**Comment by aarondav** [comment]
Looks good from my end.

**Comment by squito** [comment]
@tomwhite there is a legit failure here, looks like you need to merge w/ master and fix a compile error

**Comment by tomwhite** [comment]
I rebased this on master. Is there a way to get this to be retested?

**Comment by kayousterhout** [comment]
Jenkins, retest this please

**Comment by rxin** [comment]
I triggered Jenkins.

**Comment by squito** [comment]
@tomwhite still looks like a real compile error: 

```
[error] /home/jenkins/workspace/NewSparkPullRequestBuilder/core/src/main/scala/org/apache/spark/scheduler/DAGScheduler.scala:822: not enough arguments for method abortStage: (failedStage: org.apache.spark.scheduler.Stage, reason: String, exception: Option[Throwable])Unit.
[error] Unspecified value parameter exception.
[error]         abortStage(stage, s"Task creation failed: $e\n${e.getStackTraceString}")
[error]                   ^
```

**Comment by tomwhite** [comment]
Thanks. The error came from new code that was committed in SPARK-4352. I've rebased and fixed the offending line.

**Comment by squito** [comment]
merged to master & 1.5  thanks @tomwhite !
